src/async/pitfalls/pin.md

```diff
@@ -1,108 +1,108 @@
+change. That is why we need to `pin` futures. Using the same future repeatedly
+in a `select!` often leads to issues with pinned values.
 
 ```rust,editable,compile_fail
-use tokio::sync::mpsc::{self, Receiver};
+use tokio::sync::{mpsc, oneshot};
+use tokio::task::spawn;
 use tokio::time::{sleep, Duration};
 
-#[derive(Debug, PartialEq)]
-struct Runner {
-    name: String,
+// A work item. In this case, just sleep for the given time and respond
+// with a message on the `respond_on` channel.
+#[derive(Debug)]
+struct Work {
+    sleep_for: Duration,
+    respond_on: oneshot::Sender<u128>,
 }
 
-async fn race_finish_line(mut rcv: Receiver<String>, timeout: Duration) -> Option<Vec<Runner>> {
-    let mut performances: Vec<Runner> = Vec::new();
-    let timeout_sleep = sleep(timeout);
-    // Pinning here allows us to await `timeout_sleep` multiple times.
-    tokio::pin!(timeout_sleep);
-
-    loop {
-        tokio::select! {
-            // Rcv.recv() returns a new future every time, hence it does not need to be pinned.
-            name = rcv.recv() => performances.push(Runner { name: name? }),
-            _ = timeout_sleep.as_mut() => break
-        }
+// A worker which listens for work on a queue and performs it.
+async fn worker(mut work_queue: mpsc::Receiver<Work>) {
+    while let Some(work) = work_queue.recv().await {
+        sleep(work.sleep_for).await;
+        work.respond_on
+            .send(work.sleep_for.as_millis())
+            .expect("failed to send response");
     }
-    Some(performances)
+}
+
+// A requester which requests work and waits for it to complete.
+async fn do_work(work_queue: &mpsc::Sender<Work>, duration: Duration) -> u128 {
+    let (tx, rx) = oneshot::channel();
+    work_queue
+        .send(Work {
+            sleep_for: duration,
+            respond_on: tx,
+        })
+        .await
+        .expect("failed to send on work queue");
+    rx.await.expect("failed waiting for response")
 }
 
 #[tokio::main]
 async fn main() {
-    let (sender, receiver) = mpsc::channel(32);
-
-    let names_and_time = [
-        ("Leo", 9),("Milo", 3),("Luna", 13),("Oliver", 5),("Charlie", 11),
-    ];
-
-    let finish_line_future = race_finish_line(receiver, Duration::from_secs(6));
-
-    for (name, duration_secs) in names_and_time {
-        let sender = sender.clone();
-        tokio::spawn(async move {
-            sleep(Duration::from_secs(duration_secs)).await;
-            sender.send(String::from(name)).await.expect("Failed to send runner");
-        });
-    }
-
-    println!("{:?}", finish_line_future.await.expect("Failed to collect finish line"));
-    // [Runner { name: "Milo" }, Runner { name: "Oliver" }]
+    let (tx, rx) = mpsc::channel(10);
+    spawn(worker(rx));
+    let resp = do_work(&tx, Duration::from_millis(10)).await;
+    println!("work result: {resp}");
 }
 ```
 
+Try making the worker print the number of work items handled every 100ms.
```

Comment:
Should this go to the details part?

Reply:
This was a bit of a weird case because the example as-written does not illustrate the issue. But maybe I could save some time by finishing the first few steps in the speaker notes. I'll adjust.